src/jit/codegenarm64.cpp

```diff
@@ -5059,9 +5059,104 @@ void CodeGen::genHWIntrinsicSimdBinaryOp(GenTreeHWIntrinsic* node)
     genProduceReg(node);
 }
 
+template <typename HWIntrinsicSwitchCaseBody>
+void CodeGen::genHWIntrinsicSwitchTable(regNumber                 swReg,
+                                        regNumber                 tmpReg,
+                                        int                       swMax,
+                                        HWIntrinsicSwitchCaseBody emitSwCase)
+{
+    assert(swMax > 0);
+    assert(swMax <= 256);
+
+    assert(genIsValidIntReg(tmpReg));
+    assert(genIsValidIntReg(swReg));
+
+    BasicBlock* labelFirst       = genCreateTempLabel();
+    BasicBlock* labelBreakTarget = genCreateTempLabel();
+
+    // Detect and throw out of range exception
+    getEmitter()->emitIns_R_I(INS_cmp, EA_4BYTE, swReg, swMax);
+
+    emitJumpKind jmpGEU = genJumpKindForOper(GT_GE, CK_UNSIGNED);
+    genJumpToThrowHlpBlk(jmpGEU, SCK_ARG_RNG_EXCPN);
+
+    // Calculate switch target
+    labelFirst->bbFlags |= BBF_JMP_TARGET;
+
+    // tmpReg = labelFirst
+    getEmitter()->emitIns_R_L(INS_adr, EA_PTRSIZE, labelFirst, tmpReg);
+
+    // tmpReg = labelFirst + swReg * 8
+    getEmitter()->emitIns_R_R_R_I(INS_add, EA_PTRSIZE, tmpReg, tmpReg, swReg, 3, INS_OPTS_LSL);
+
+    // br tmpReg
+    getEmitter()->emitIns_R(INS_br, EA_PTRSIZE, tmpReg);
+
+    genDefineTempLabel(labelFirst);
+    for (int i = 0; i < swMax; ++i)
+    {
+        emitSwCase(i);
+        inst_JMP(EJ_jmp, labelBreakTarget);
+    }
+    genDefineTempLabel(labelBreakTarget);
+}
```

Comment:
Why hardcode it, instead of defining a label per offset, so it can be dynamically sized, if required?

That is what I did for the x86 codegen (https://github.com/dotnet/coreclr/blob/master/src/jit/hwintrinsiccodegenxarch.cpp#L655).

I do like the approach to making it reusable 😄

Reply:
> Why hardcode it, instead of defining a label per offset, so it can be dynamically sized, if required?

Arm64 instructions are fixed size.  So it makes sense.  Single instruction per case is the expected use case.  (I would add a case instruction count parameter if needed).

Label per offset would require more complexity.  Either a separate direct branch table so the target address could be calculated, or an if else compare and branch chain.  Or ...

Reply:
Looks like you chose the separate direct branch table.

Reply:
The generated code will be smaller.  Since the jump table is not needed.

Certainly if a variable sized case was needed, We could add the jump table.  Right now I expect homogeneous cases of single instructions.

Reply:
> Looks like you chose the separate direct branch table

Yes, x86 instructions can potentially range in size from 1 byte to 16+ bytes (depending on prefixes, encoding, etc).

Although, within a given HWIntrinsic jump table, they will likely be fairly consistent.

Reply:
It seems that for Arm64 this approach is best. If it were found to be necessary, it could be parameterized based on the instruction (though I don't see us needing this for anything that would require more than a single instruction per case).